src/varnish.c

```diff
@@ -127,36 +134,106 @@ static int varnish_submit_gauge(const char *plugin_instance, /* {{{ */
                                 const char *category, const char *type,
                                 const char *type_instance,
                                 uint64_t gauge_value) {
-  return varnish_submit(plugin_instance, category, type, type_instance,
+  return varnish_submit(plugin_instance, category, NULL, type, type_instance,
                         (value_t){
                             .gauge = (gauge_t)gauge_value,
                         });
 } /* }}} int varnish_submit_gauge */
 
+static int
+varnish_submit_gauge_with_target(const char *plugin_instance, /* {{{ */
+                                 const char *category, const char *target,
+                                 const char *type, const char *type_instance,
+                                 uint64_t gauge_value) {
+
+  return varnish_submit(plugin_instance, category, target, type, type_instance,
+                        (value_t){
+                            .gauge = (gauge_t)gauge_value,
+                        });
+} /* }}} int varnish_submit_gauge_with_target */
+
 static int varnish_submit_derive(const char *plugin_instance, /* {{{ */
                                  const char *category, const char *type,
                                  const char *type_instance,
                                  uint64_t derive_value) {
-  return varnish_submit(plugin_instance, category, type, type_instance,
+  return varnish_submit(plugin_instance, category, NULL, type, type_instance,
                         (value_t){
                             .derive = (derive_t)derive_value,
                         });
 } /* }}} int varnish_submit_derive */
 
+static int
+varnish_submit_derive_with_target(const char *plugin_instance, /* {{{ */
+                                  const char *category, const char *target,
+                                  const char *type, const char *type_instance,
+                                  uint64_t derive_value) {
+  return varnish_submit(plugin_instance, category, target, type, type_instance,
+                        (value_t){
+                            .derive = (derive_t)derive_value,
+                        });
+} /* }}} int varnish_submit_derive_with_target */
+
 #if HAVE_VARNISH_V3 || HAVE_VARNISH_V4 || HAVE_VARNISH_V5 || HAVE_VARNISH_V6
 static int varnish_monitor(void *priv,
                            const struct VSC_point *const pt) /* {{{ */
 {
   uint64_t val;
   const user_config_t *conf;
   const char *name;
+  //  const char *stat_category;
+  const char *stat_target;
 
   if (pt == NULL)
     return 0;
 
   conf = priv;
 
-#if HAVE_VARNISH_V5 || HAVE_VARNISH_V6
+#if HAVE_VARNISH_V6
+  /*
+   stats examples:
+   MAIN.threads
+   SMA.s0.c_req
+   SMA.Transient.c_req
+   VBE.vclName.backendName.req // where vclName changes every time a new vcl is
+   loaded
+  */
+  const int MAX_TOKENS = 4; // we expect 4 maximum tokens
+  char *token_array[MAX_TOKENS];
+  int i = 0;
```

Comment:
I'm trying to split the string by dot, and get the last token as name, and previous token as target. but I also need to know how many tokens were there to know if target exists (and is not the first token which is category).

At first I also wanted to check for category, so was trying to get the full token array, but since I'm not checking for it, this is overkill.

So v5 code might work to get the two last tokens, but I'll need to count tokens as well.

Reply:
How about this:

```c
char *tokens[4] = {NULL};
size_t tokens_num = 0;
char *ptr = buffer;
char *saveptr = NULL;
while ((token = strtok_r(ptr, ".", &saveptr)) != NULL) {
  ptr = NULL;
  if (tokens_num < STATIC_ARRAY_SIZE(tokens)) {
    tokens[tokens_num] = token;
  }
  tokens_num++;
}
if ((tokens_num < 2) || (tokens_num > STATIC_ARRAY_SIZE(tokens))) {
  // handle error
}

// Note how all these assignments are close to each other,
// making it easy to grasp what is going on.
// "array[len - 1]" is a very common pattern for "give me the last element",
// "array[len - 2]" is also easy to read in this context.
category = tokens[0];
name = tokens[tokens_num - 1];
if (tokens_num >= 3) {
  target = tokens[tokens_num - 2];
}
```

*Caveat: typed from memory, mistakes likely*

Reply:
Thanks a lot, all assignments are close to each other, but since i need to copy strings, it's not as compact as your solution :)